plugins/ltac2/g_ltac2.mlg

```diff
@@ -798,7 +798,21 @@ GRAMMAR EXTEND Gram
     [ [ m = match_list -> { m } ] ]
   ;
   gmatch_hyp_pattern:
-    [ [ na = Prim.name; ":"; pat = match_pattern -> { (na, pat) } ] ]
+    [ [ na = Prim.name; ":"; pat = match_pattern -> { (na, None, pat) }
+      | na = Prim.name; ":="; "["; bod = match_pattern; "]"; ":"; pat = match_pattern ->
+        { (na, Some bod, pat) }
+      | na = Prim.name; ":="; bod = match_pattern ->
+        { let bod, ty = match bod.CAst.v with
+              | QConstrMatchPattern { CAst.v = CCast (t, _, ty) } ->
```

Comment:
Not sure if we care, but I think this leads to slightly surprising behavior in treating VM and native casts (which I guess is already present in Ltac1?), namely that they'll match terms without casts without a warning (whereas in most places there's a warning that casts are ignored in matching).  I think the most correct behavior would be to add a warning when the cast kind is anything other than the default cast kind, both here and in Ltac1.  This shouldn't block this PR though.

Reply:
The idea of this code is to make `na := foo : ty` by considering it more intuitive that the user intended to have `ty` match the type of the hypothesis rather than matching a cast in the body. So IMO the right thing to do is to only match defaultcast. I'll push that change and make a separate PR for ltac1.